// @wdio/jasmine-framework/build/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.adapterFactory = exports.JasmineAdapter = void 0;
const jasmine_1 = __importDefault(require("jasmine"));
const utils_1 = require("@wdio/utils");
const logger_1 = __importDefault(require("@wdio/logger"));
const reporter_1 = __importDefault(require("./reporter"));
const INTERFACES = {
    bdd: ['beforeAll', 'beforeEach', 'it', 'xit', 'fit', 'afterEach', 'afterAll']
};
const TEST_INTERFACES = ['it', 'fit', 'xit'];
const NOOP = function noop() { };
const DEFAULT_TIMEOUT_INTERVAL = 60000;
const log = logger_1.default('@wdio/jasmine-framework');
class JasmineAdapter {
    constructor(_cid, _config, _specs, _capabilities, reporter) {
        this._cid = _cid;
        this._config = _config;
        this._specs = _specs;
        this._capabilities = _capabilities;
        this._totalTests = 0;
        this._hookIds = 0;
        this._hasTests = true;
        this._jasmineNodeOpts = Object.assign({
            cleanStack: true
        }, this._config.jasmineNodeOpts);
        this._reporter = new reporter_1.default(reporter, {
            cid: this._cid,
            specs: this._specs,
            cleanStack: this._jasmineNodeOpts.cleanStack
        });
        this._hasTests = true;
    }
    async init() {
        const self = this;
        this._jrunner = new jasmine_1.default({});
        const { jasmine } = this._jrunner;
        const jasmineEnv = jasmine.getEnv();
        this._jrunner.projectBaseDir = '';
        this._jrunner.specDir = '';
        this._jrunner.addSpecFiles(this._specs);
        jasmine.DEFAULT_TIMEOUT_INTERVAL = this._jasmineNodeOpts.defaultTimeoutInterval || DEFAULT_TIMEOUT_INTERVAL;
        jasmineEnv.addReporter(this._reporter);
        jasmineEnv.configure({
            specFilter: this._jasmineNodeOpts.specFilter || this.customSpecFilter.bind(this),
            stopOnSpecFailure: Boolean(this._jasmineNodeOpts.stopOnSpecFailure),
            failSpecWithNoExpectations: Boolean(this._jasmineNodeOpts.failSpecWithNoExpectations),
            failFast: this._jasmineNodeOpts.failFast,
            random: Boolean(this._jasmineNodeOpts.random),
            seed: Boolean(this._jasmineNodeOpts.seed),
            oneFailurePerSpec: Boolean(this._jasmineNodeOpts.stopSpecOnExpectationFailure ||
                this._jasmineNodeOpts.oneFailurePerSpec)
        });
        jasmine.Spec.prototype.addExpectationResult = this.getExpectationResultHandler(jasmine);
        const hookArgsFn = (context) => [{ ...(self._lastTest || {}) }, context];
        const emitHookEvent = (fnName, eventType) => (_test, _context, { error } = {}) => {
            const title = `"${fnName === 'beforeAll' ? 'before' : 'after'} all" hook`;
            const hook = {
                id: '',
                start: new Date(),
                type: 'hook',
                description: title,
                fullName: title,
                ...(error ? { error } : {})
            };
            this._reporter.emit('hook:' + eventType, hook);
        };
        INTERFACES['bdd'].forEach((fnName) => {
            const isTest = TEST_INTERFACES.includes(fnName);
            const beforeHook = [...this._config.beforeHook];
            const afterHook = [...this._config.afterHook];
            if (fnName.includes('All')) {
                beforeHook.push(emitHookEvent(fnName, 'start'));
                afterHook.push(emitHookEvent(fnName, 'end'));
            }
            this._config.beforeTest;
            utils_1.runTestInFiberContext(isTest, isTest ? this._config.beforeTest : beforeHook, hookArgsFn, isTest ? this._config.afterTest : afterHook, hookArgsFn, fnName, this._cid);
        });
        jasmine_1.default.prototype.configureDefaultReporter = NOOP;
        let beforeAllMock = jasmine.Suite.prototype.beforeAll;
        jasmine.Suite.prototype.beforeAll = function (...args) {
            self._lastSpec = this.result;
            beforeAllMock.apply(this, args);
        };
        let executeMock = jasmine.Spec.prototype.execute;
        jasmine.Spec.prototype.execute = function (...args) {
            self._lastTest = this.result;
            self._lastTest.start = new Date().getTime();
            executeMock.apply(this, args);
        };
        this._loadFiles();
        const { setOptions } = require('expect-webdriverio');
        setOptions({
            wait: this._config.waitforTimeout,
            interval: this._config.waitforInterval,
        });
        return this;
    }
    _loadFiles() {
        if (!this._jrunner) {
            throw new Error('Jasmine not initiate yet');
        }
        try {
            if (Array.isArray(this._jasmineNodeOpts.requires)) {
                this._jrunner.addRequires(this._jasmineNodeOpts.requires);
            }
            if (Array.isArray(this._jasmineNodeOpts.helpers)) {
                this._jrunner.addHelperFiles(this._jasmineNodeOpts.helpers);
            }
            this._jrunner.loadRequires();
            this._jrunner.loadHelpers();
            this._jrunner.loadSpecs();
            this._grep(this._jrunner.env.topSuite());
            this._hasTests = this._totalTests > 0;
        }
        catch (err) {
            log.warn('Unable to load spec files quite likely because they rely on `browser` object that is not fully initialised.\n' +
                '`browser` object has only `capabilities` and some flags like `isMobile`.\n' +
                'Helper files that use other `browser` commands have to be moved to `before` hook.\n' +
                `Spec file(s): ${this._specs.join(',')}\n`, 'Error: ', err);
        }
    }
    _grep(suite) {
        suite.children.forEach((child) => {
            if (Array.isArray(child.children)) {
                return this._grep(child);
            }
            if (this.customSpecFilter(child)) {
                this._totalTests++;
            }
        });
    }
    hasTests() {
        return this._hasTests;
    }
    async run() {
        const result = await new Promise((resolve, reject) => {
            if (!this._jrunner) {
                return reject(new Error('Jasmine not initiate yet'));
            }
            this._jrunner.env.beforeAll(this.wrapHook('beforeSuite'));
            this._jrunner.env.afterAll(this.wrapHook('afterSuite'));
            this._jrunner.onComplete(() => resolve(this._reporter.getFailedCount()));
            this._jrunner.execute();
        });
        await utils_1.executeHooksWithArgs('after', this._config.after, [result, this._capabilities, this._specs]);
        return result;
    }
    customSpecFilter(spec) {
        const { grep, invertGrep } = this._jasmineNodeOpts;
        const grepMatch = !grep || spec.getFullName().match(new RegExp(grep)) !== null;
        if (grepMatch === Boolean(invertGrep)) {
            spec.pend('grep');
            return false;
        }
        return true;
    }
    wrapHook(hookName) {
        return (done) => utils_1.executeHooksWithArgs(hookName, this._config[hookName], [this.prepareMessage(hookName)]).then(() => done(), (e) => {
            log.info(`Error in ${hookName} hook: ${e.stack.slice(7)}`);
            done();
        });
    }
    prepareMessage(hookName) {
        var _a, _b;
        const params = { type: hookName };
        switch (hookName) {
            case 'beforeSuite':
            case 'afterSuite':
                params.payload = Object.assign({
                    file: (_a = this._jrunner) === null || _a === void 0 ? void 0 : _a.specFiles[0]
                }, this._lastSpec);
                break;
            case 'beforeTest':
            case 'afterTest':
                params.payload = Object.assign({
                    file: (_b = this._jrunner) === null || _b === void 0 ? void 0 : _b.specFiles[0]
                }, this._lastTest);
                break;
        }
        return this.formatMessage(params);
    }
    formatMessage(params) {
        var _a;
        let message = {
            type: params.type
        };
        if (params.payload) {
            message.title = params.payload.description;
            message.fullName = params.payload.fullName || null;
            message.file = params.payload.file;
            if (params.payload.failedExpectations && params.payload.failedExpectations.length) {
                message.errors = params.payload.failedExpectations;
                message.error = params.payload.failedExpectations[0];
            }
            if (params.payload.id && params.payload.id.startsWith('spec')) {
                message.parent = (_a = this._lastSpec) === null || _a === void 0 ? void 0 : _a.description;
                message.passed = params.payload.failedExpectations.length === 0;
            }
            if (params.type === 'afterTest') {
                message.duration = new Date().getTime() - params.payload.start;
            }
            if (typeof params.payload.duration === 'number') {
                message.duration = params.payload.duration;
            }
        }
        return message;
    }
    getExpectationResultHandler(jasmine) {
        let { expectationResultHandler } = this._jasmineNodeOpts;
        const origHandler = jasmine.Spec.prototype.addExpectationResult;
        if (typeof expectationResultHandler !== 'function') {
            return origHandler;
        }
        return this.expectationResultHandler(origHandler);
    }
    expectationResultHandler(origHandler) {
        const { expectationResultHandler } = this._jasmineNodeOpts;
        return function (passed, data) {
            try {
                expectationResultHandler.call(this, passed, data);
            }
            catch (e) {
                if (passed) {
                    passed = false;
                    data = {
                        passed,
                        message: 'expectationResultHandlerError: ' + e.message,
                        error: e
                    };
                }
            }
            return origHandler.call(this, passed, data);
        };
    }
}
exports.JasmineAdapter = JasmineAdapter;
const adapterFactory = {};
exports.adapterFactory = adapterFactory;
adapterFactory.init = async function (...args) {
    const adapter = new JasmineAdapter(...args);
    const instance = await adapter.init();
    return instance;
};
exports.default = adapterFactory;
